cmd/titus-log-uploader/main.go

```diff
@@ -0,0 +1,113 @@
+package main
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/Netflix/metrics-client-go/metrics"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Netflix/titus-executor/filesystems"
+	kubeviewer "github.com/Netflix/titus-executor/logviewer/kubelet"
+	"github.com/Netflix/titus-executor/tag"
+	"github.com/Netflix/titus-executor/uploader"
+)
+
+/*
+	This binary handles log persistence and access from within a pod.
+
+	It accomplishes this by having an injected shared volume and a webserver.  The binary monitors this shared volume and publishes to s3 on a specified cadience.  Additionally it provides a simple web server for stating and listing these logs in accordance with the logviewer API from the titus-api server.
+*/
+
+func main() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	m := metrics.New(ctx, logrus.StandardLogger(), tag.Defaults)
+
+	wconf, err := watchConfigFromEnvironment()
+	if err != nil {
+		logrus.Fatal(err)
+	}
+
+	uc, err := uploadConfigFromEnvironment()
+	if err != nil {
+		logrus.Fatal(err)
+	}
+
+	lc, err := logViewerConfigFromEnvironment()
+	if err != nil {
+		logrus.Fatal(err)
+	}
+
+	logrus.WithFields(logrus.Fields{
+		"watchConfig":  wconf,
+		"uploadConfig": uc,
+		"viewerConfig": lc,
+	}).Info("Booting with config")
+
+	logviewer := &kubeviewer.VolumeLogViewer{
+		Volume: lc.Volume,
+	}
+
+	server := http.Server{
+		Addr:         ":8004", // TODO[cconger]: Config this port
+		Handler:      logviewer.AttachHandlers(&http.ServeMux{}),
+		ReadTimeout:  10 * time.Second,
+		WriteTimeout: 10 * time.Second,
+	}
+
+	s3Backend, err := uploader.NewS3Backend(m, uc.bucketName, uc.pathPrefix, uc.taskRole, uc.taskID, uc.writerRole, uc.useDefaultRole)
+	if err != nil {
+		logrus.Fatal(err)
+	}
+	u := uploader.NewUploaderWithBackend(s3Backend)
+
+	watcher, err := filesystems.NewWatcher(m, wconf, u)
+	if err != nil {
+		logrus.Fatal(err)
+	}
+
+	err = watcher.Watch(ctx)
+	if err != nil {
+		logrus.Fatal(err)
+	}
+
+	// Start the logviewer webserver
```

Comment:
My preference is to make this a different binary, and inject separate uploading and viewing sidecars.

Reply:
I can do that, I'd appreciate you fleshing out your thinking on that, would you ever want one without the other?

Reply:
Yes, exactly. We already have customers who run filebeat inside their containers (and some day their pods).

Reply:
Didn't @rgulewich already build that?

Reply:
We use the same logviewer that BaseOS installs on all the EC2 instances, packaged up as a container.  It implements all of the old Titus "legacy" endpoints for backward compatibility.

This gets us out of the business of building a second logviewer, and keeps it consistent between EC2 and Titus.

Reply:
The implementation in `cmd/titus-logviewer/...` is the old logviewer. It's kept around so that the integration tests in this repo have something to test against, but isn't used otherwise.